torch/csrc/autograd/record_function.h

```diff
@@ -262,52 +347,100 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
+//       callbacks specific to the given thread
+//  - we allow the added callbacks to be sampled, by specifying a sampling
+//    probability for each callback pair, if the start callback is
+//    not picked to run, the corresponding end callback won't be called
+//  - a typical use case for the global callbacks is passive monitoring
+//    in the background (e.g. fleet-wide monitoring), without focusing on
+//    the specific peice of code
+//  - in contrast, thread local callbacks are enabled locally, on demand,
+//    for the specific piece of code (range) and are not sampled
+//  - a typical use case for thread local callbacks is profiler and code
+//    execution tracer
+//     - note, some functionality (e.g. profiler) can automatically
+//       propagate its calbacks across thread by using ThreadLocalState
+//       mechanism, but in general callbacks are not propagated
+//  - adding/removing global callbacks is not thread safe and should be done
+//    only when no other code is running, e.g. during the initialization
+
+/**
+ * addThreadLocalCallback adds a thread local callback to run with RecordFunction,
+ * returns handle to use with removeThreadLocalCallback
+ */
+TORCH_API uint64_t addThreadLocalCallback(RecordFunctionCallback cb);
+
+/**
+ * hasThreadLocalCallbacks returns whether there're callbacks registered
+ * with addThreadLocalCallback
+ */
+TORCH_API bool hasThreadLocalCallbacks();
+
 /**
- * pushCallback adds a pair of callbacks to run with RecordFunction:
- *  start, end - the callbacks to run when entering and exiting the scope;
- *    if start callback returns false, end callback won't be executed;
- *  needs_inputs - whether the callbacks need the inputs passed from the observed
- *    function/range; NOTE: passing the inputs incurs an additional overhead;
- *  sampling_prob - whether the callbacks are sampled and the sampling
- *    probability;
- *  scopes - types of scopes to execute the callbacks on (see RecordScope);
- *    passing empty set means the callbacks will be executed for all possible
- *    scope types
+ * clearThreadLocalCallbacks removes all thread local callbacks
+ */
+TORCH_API void clearThreadLocalCallbacks();
+
+/**
+ * addGlobalCallback adds a global callback to run with RecordFunction:
  *
- * WARNING: not thread safe, must not overlap with other PyTorch code execution
+ * WARNING: not thread safe, typically addGlobalCallback can be called
+ * only during the program initialization
  */
-TORCH_API void pushCallback(
-    std::function<bool(const RecordFunction&)> start,
-    std::function<void(const RecordFunction&)> end =
-        [](const RecordFunction&) {},
-    bool needs_inputs = false,
-    double sampling_prob = 1.0,
-    std::unordered_set<RecordScope, std::hash<RecordScope>> scopes =
-        std::unordered_set<RecordScope, std::hash<RecordScope>>());
+TORCH_API uint64_t addGlobalCallback(RecordFunctionCallback cb);
```

Comment:
maybe be a bit typesafe and return a wrapper struct for the handle (like c10 registrations do)

Reply:
maybe I could go with typedef first,
If we need to generalize, I can then replace typedef with actual struct